File: implementations/sdk/php/src/SchemaValidator.php
<?php

declare(strict_types=1);

namespace Lcp;

use Opis\JsonSchema\CompliantValidator;
use RuntimeException;

final class SchemaValidator
{
    private CompliantValidator $validator;
    /** @var array<string, string> */
    private array $idsByName = [];
    /** @var array<string, array<string, mixed>> */
    private array $documents = [];

    /** @param array<string, string> $schemaTexts */
    public function __construct(array $schemaTexts)
    {
        $this->validator = new CompliantValidator();
        $resolver = $this->validator->resolver();
        foreach ($schemaTexts as $name => $text) {
            $schema = json_decode($text, true, 512, JSON_THROW_ON_ERROR);
            $id = is_array($schema) && isset($schema['$id']) ? (string)$schema['$id'] : $name;
            $this->documents[$this->normalize($name)] = $schema;
            $this->documents[$this->normalize($id)] = $schema;
            $resolver->registerRaw($text, $id);
            $this->idsByName[$this->normalize($name)] = $id;
            $this->idsByName[$this->normalize($id)] = $id;
        }
    }

    public static function fromDirectory(string $root): self
    {
        $schemas = [];
        $iterator = new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($root));
        foreach ($iterator as $file) {
            if (!$file->isFile() || $file->getExtension() !== 'json') continue;
            $schemas[str_replace('\\', '/', substr($file->getPathname(), strlen(rtrim($root, DIRECTORY_SEPARATOR)) + 1))] = file_get_contents($file->getPathname());
        }
        if (basename(rtrim($root, DIRECTORY_SEPARATOR)) === 'schemas') {
            $verticalRoot = dirname(rtrim($root, DIRECTORY_SEPARATOR)) . DIRECTORY_SEPARATOR . 'verticals';
            if (is_dir($verticalRoot)) {
                $iterator = new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($verticalRoot));
                foreach ($iterator as $file) {
                    if (!$file->isFile() || $file->getExtension() !== 'json') continue;
                    $relative = str_replace('\\', '/', substr($file->getPathname(), strlen(rtrim($verticalRoot, DIRECTORY_SEPARATOR)) + 1));
                    $schemas['verticals/' . $relative] = file_get_contents($file->getPathname());
                }
            }
        }
        return new self($schemas);
    }

    public function validate(string $schemaName, mixed $document): void
    {
        $id = $this->idsByName[$this->normalize($schemaName)] ?? null;
        if ($id === null) throw new RuntimeException("Unknown LCP schema: $schemaName");
        $result = $this->validator->validate($this->jsonValue($document), $id);
        if (!$result->isValid()) throw new RuntimeException("LCP schema validation failed for $schemaName");
    }

    public function validateEnvelope(array $envelope): void
    {
        $this->validate('schemas/envelope.json', $envelope);
        $type = $envelope['lcp']['message']['type'] ?? null;
        if (!is_string($type)) throw new RuntimeException('Missing lcp.message.type');
        $payload = $envelope['lcp']['payload'] ?? null;
        $this->validate("schemas/$type.json", $payload);
        $this->validateVerticalPolicy($type, $payload);
    }

    private function validateVerticalPolicy(string $type, mixed $payload): void
    {
        if (!in_array($type, ['lead', 'call', 'post', 'ping'], true) || !is_array($payload)) return;
        $attributes = $payload['attributes'] ?? null;
        if (!is_array($attributes)) return;
        $vertical = $type === 'ping' ? ($payload['vertical'] ?? null) : ($attributes['vertical'] ?? null);
        if (!is_string($vertical) || $vertical === '') return;
        $schema = $this->documents[$this->normalize("verticals/$vertical.json")] ?? null;
        if (!is_array($schema)) throw new RuntimeException("Vertical schema '$vertical' not found");
        $verticalAttributes = $attributes;
        if ($type === 'ping') {
            $verticalAttributes['vertical'] ??= $vertical;
            $verticalAttributes['schema_version'] ??= $schema['properties']['schema_version']['const'] ?? '1.0.0';
        }
        $this->validateVertical($vertical, $verticalAttributes);
        if ($type === 'ping') $this->validatePingSafe($vertical, $attributes, $schema, 'attributes');
    }

    /** @param array<string, mixed> $value */
    private function validatePingSafe(string $vertical, array $value, array $schema, string $path): void
    {
        $properties = $schema['properties'] ?? [];
        foreach ($value as $name => $child) {
            if ($path === 'attributes' && in_array($name, ['vertical', 'schema_version'], true)) continue;
            $definition = $properties[$name] ?? null;
            if (!is_array($definition) || ($definition['ping_safe'] ?? false) !== true) {
                throw new RuntimeException("$path.$name is not tagged ping_safe: true in vertical '$vertical'");
            }
            if (is_array($child) && isset($definition['properties']) && is_array($definition['properties'])) {
                $this->validatePingSafe($vertical, $child, $definition, "$path.$name");
            }
        }
    }

    public function validateOffer(array $offer): void { $this->validate('schemas/offer.json', $offer); }
    public function validateVertical(string $vertical, array $attributes): void { $this->validate("verticals/$vertical.json", $attributes); }

    private function jsonValue(mixed $value): mixed
    {
        if (is_array($value)) {
            $converted = [];
            foreach ($value as $key => $item) $converted[$key] = $this->jsonValue($item);
            return array_is_list($value) ? $converted : (object)$converted;
        }
        return $value;
    }

    private function normalize(string $name): string
    {
        $name = ltrim(str_replace('\\', '/', $name), '/');
        $name = preg_replace('/^(schemas|verticals)\//', '', $name) ?? $name;
        return preg_replace('/\.json$/', '', $name) ?? $name;
    }
}

File: implementations/sdk/ruby/lib/schema_validator.rb
# frozen_string_literal: true

require "json"

module LcpSdk
  class SchemaValidationError < StandardError; end

  class SchemaValidator
    def initialize(schema_documents)
      require "json_schemer"
      # Keep a private copy so self-references in core.json can be made
      # fragment-local without mutating the caller's canonical documents.
      @documents = schema_documents.transform_values do |document|
        JSON.parse(JSON.generate(document))
      end
      @documents.each_value do |document|
        rewrite_core_refs(document) if document["$id"] == "https://lcp.dev/schemas/core.json"
      end
      @ids = {}
      @documents.each do |name, document|
        id = document.fetch("$id", name)
        @ids[normalize(name)] = id
        @ids[normalize(id)] = id
      end
    end

    def self.from_directory(root)
      documents = {}
      Dir.glob(File.join(root, "**", "*.json")).each do |path|
        name = path.delete_prefix("#{root}/")
        documents[name] = JSON.parse(File.read(path))
      end
      if File.basename(root.to_s) == "schemas"
        vertical_root = File.join(File.dirname(root), "verticals")
        Dir.glob(File.join(vertical_root, "**", "*.json")).each do |path|
          name = path.delete_prefix("#{vertical_root}/")
          documents["verticals/#{name}"] = JSON.parse(File.read(path))
        end
      end
      new(documents)
    end

    def validate(schema_name, instance)
      id = @ids.fetch(normalize(schema_name)) { raise KeyError, "unknown LCP schema: #{schema_name}" }
      schema = @documents.values.find { |candidate| candidate["$id"] == id } || @documents.fetch(schema_name)
      errors = JSONSchemer.schema(schema, ref_resolver: lambda do |uri|
        reference = uri.to_s
        @documents.values.find { |candidate| candidate["$id"].to_s == reference } ||
          @documents[reference] ||
          @documents[reference.sub(%r{.*/}, "")]
      end).validate(instance).to_a
      raise SchemaValidationError, "LCP schema validation failed for #{schema_name}: #{errors.first}" unless errors.empty?
      true
    end

    def validate_envelope(envelope)
      validate("schemas/envelope.json", envelope)
      type = envelope.dig("lcp", "message", "type")
      raise SchemaValidationError, "missing lcp.message.type" unless type.is_a?(String)
      payload = envelope.dig("lcp", "payload")
      validate("schemas/#{type}.json", payload)
      validate_vertical_policy(type, payload)
    end

    def validate_vertical_policy(type, payload)
      return unless %w[lead call post ping].include?(type) && payload.is_a?(Hash)
      attributes = payload["attributes"]
      return unless attributes.is_a?(Hash)
      vertical = type == "ping" ? payload["vertical"] : attributes["vertical"]
      return unless vertical.is_a?(String) && !vertical.empty?
      schema = @documents[normalize("verticals/#{vertical}.json")] ||
        @documents["verticals/#{vertical}.json"] ||
        @documents["#{vertical}.json"]
      raise SchemaValidationError, "Vertical schema '#{vertical}' not found" unless schema.is_a?(Hash)
      vertical_attributes = JSON.parse(JSON.generate(attributes))
      if type == "ping"
        vertical_attributes["vertical"] ||= vertical
        vertical_attributes["schema_version"] ||= schema.dig("properties", "schema_version", "const") || "1.0.0"
      end
      validate_vertical(vertical, vertical_attributes)
      validate_ping_safe(vertical, attributes, schema, "attributes") if type == "ping"
    end

    def validate_ping_safe(vertical, value, schema, path)
      return unless value.is_a?(Hash)
      properties = schema.fetch("properties", {})
      value.each do |name, child|
        next if path == "attributes" && %w[vertical schema_version].include?(name)
        definition = properties[name]
        unless definition.is_a?(Hash) && definition["ping_safe"] == true
          raise SchemaValidationError, "#{path}.#{name} is not tagged ping_safe: true in vertical '#{vertical}'"
        end
        validate_ping_safe(vertical, child, definition, "#{path}.#{name}") if child.is_a?(Hash) && definition["properties"].is_a?(Hash)
      end
    end

    def validate_offer(offer)
      validate("schemas/offer.json", offer)
    end

    def validate_vertical(vertical, attributes)
      validate("verticals/#{vertical}.json", attributes)
    end

    private

    def rewrite_core_refs(value)
      case value
      when Hash
        if value["$ref"].is_a?(String)
          value["$ref"] = value["$ref"].sub(/\Acore\.json(?=#)/, "")
        end
        value.each_value { |child| rewrite_core_refs(child) }
      when Array
        value.each { |child| rewrite_core_refs(child) }
      end
    end

    def normalize(name)
      name.to_s.tr("\\", "/").sub(%r{^/}, "").sub(%r{^(schemas|verticals)/}, "").sub(/\.json$/, "")
    end
  end
end
